[WIP] Symbolic rotation matrix

mitiguy opened the pull request.

mitiguy commented:
Feature review +@amcastro-tri 

---

Review status: all discussions resolved, platform LGTM missing

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037#-:-LFdjqzn722f5QUTnjBz:bdi8jt8)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 4 of 7 files at r1, 6 of 6 files at r2.
Review status: 6 unresolved discussions, LGTM missing from assignee amcastro-tri, platform LGTM missing

---

*[math/rotation_matrix.h, line 34 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB3KW4DL83ySaZjMez:-LGB3KW5E-IBrCGD4Wv3:b-x93z4f) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/rotation_matrix.h#L34)):*
> ```objc
> /// @note In debug builds, several methods in this class do a validity check
> /// and throw an exception (std::logic_error) if the rotation matrix is invalid.
> /// For speed in release builds, nearly all validity checks are skipped.
> ```

is this true for construction and/or factories? maybe specify?

---

*[math/rotation_matrix.h, line 420 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB3lHlB5JT_jFbFuBC:-LGB3lHlB5JT_jFbFuBD:bsz8tce) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/rotation_matrix.h#L420)):*
> ```objc
>   /// Engineering and Computer Science, Massachusetts Institute of Technology"
>   /// https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-241j-dynamic-systems-and-control-spring-2011/readings/MIT6_241JS11_chap04.pdf
>   // @internal This function is not generated for symbolic Expression.
> ```

this is not true anymore? 
And, do the `@internal` get Doxygen generated if placed within normal comments? (that is `//` instead of `///`)

---

*[math/test/rotation_conversion_test.cc, line 360 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB4smN72f1ZGfrlkxx:-LGB4smN72f1ZGfrlkxy:bb4zh49) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_conversion_test.cc#L360)):*
> ```cpp
>     // Ensure the calculated quaternion produces the same rotation matrix.
>     const RotationMatrix<double> rotmat(quat_drake);
>     EXPECT_TRUE(Ri.IsNearlyEqualTo(rotmat, 256 * kEpsilon).value());
> ```

any reason why these power of two values multiplying `kEpsilon`? maybe worth documenting so I can learn something useful :-).
also, why two orders of magnitude bigger than epsilon? a comment would help.

---

*[math/test/rotation_matrix_test.cc, line 415 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB1RAW80OoDE2uBnzD:-LGB1RAW80OoDE2uBnzE:b-tttgm7) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L415)):*
> ```cpp
> 
>   // R * R_inverse does not natually simplify to the identity matrix without
>   // a simplifying rule such as q0^2 + q1^2 + q2^2 + q3^2 = 1.
> ```

@soonho-tri, is it possible to somehow include this rule?

---

*[math/test/rotation_matrix_test.cc, line 424 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB2MSH5vDAqAQGk5HC:-LGB2MSH5vDAqAQGk5HD:b-pja24d) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L424)):*
> ```cpp
>   // Evaluate the symbolic R_R_inverse matrix with specific numerical values.
>   const Matrix3<symbolic::Expression> m_symbolic = R_R_inverse.matrix();
>   Eigen::Matrix3d m_numerical;
> ```

once #9073 merges, you'd be able to do `const Eigen::Matrix3d m_numerical{Evaluate(m_symbolic, env)};`

---

*[math/test/rotation_matrix_test.cc, line 442 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB33zr5EW6CcxRcRWT:-LGB33zr5EW6CcxRcRWU:b-l5h1r) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L442)):*
> ```cpp
>   // Verify Bool method IsOrthonormal();
>   const Bool<double> is_orthonormal =
>       RotationMatrix<double>::IsOrthonormal(R_identity.matrix(), 800*kEpsilon);
> ```

any thoughts/comments on why `800`? just from running the test or is there an order of magnitude calculation you are doing?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Sorry for the delay @mitiguy. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 6 unresolved discussions, platform LGTM missing

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037#-:-LGB5Fs35Bw-eI5wfqhS:b-sytwv7)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@seancurtis-tri for platform review please.

---

Review status: 6 unresolved discussions, platform LGTM missing

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037#-:-LGB5JhP7CM8ccpOACPF:b-tnzbcp)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I'm going to stall until after you bring this up to master.

---

Review status: 6 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037#-:-LGBHyrZ0LXwsIlhmHLN:b5j6ubs)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 6 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing

---

*[math/test/rotation_matrix_test.cc, line 424 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB2MSH5vDAqAQGk5HC:-LGBLKU79dNPYndBJrOl:bdls6zi) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L424)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

once #9073 merges, you'd be able to do `const Eigen::Matrix3d m_numerical{Evaluate(m_symbolic, env)};`
</blockquote></details>

FYI, it's merged.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 6 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing

---

*[math/test/rotation_matrix_test.cc, line 415 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB1RAW80OoDE2uBnzD:-LGCva0u9OAYfM4IYSd7:b-nwuokg) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L415)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

@soonho-tri, is it possible to somehow include this rule?
</blockquote></details>

Yes, it's doable. I'll add it to my test cases later.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 7 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing

---

*[math/test/rotation_matrix_test.cc, line 414 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGD0w3N0LOBAV9NokOw:-LGD0w3N0LOBAV9NokOx:b-u6uqwv) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L414)):*
> ```cpp
>   const RotationMatrix<symbolic::Expression> R_R_inverse = R * R_inverse;
> 
>   // R * R_inverse does not natually simplify to the identity matrix without
> ```

Nit: natually -> naturally

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
Review status: 6 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing

---

*[math/rotation_matrix.h, line 34 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB3KW4DL83ySaZjMez:-LGD1d8W9lZgvhE88g7T:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/rotation_matrix.h#L34)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is this true for construction and/or factories? maybe specify?
</blockquote></details>

Done.

---

*[math/rotation_matrix.h, line 420 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB3lHlB5JT_jFbFuBC:-LGD2YXpCAxvvZbtYaax:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/rotation_matrix.h#L420)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

this is not true anymore? 
And, do the `@internal` get Doxygen generated if placed within normal comments? (that is `//` instead of `///`)
</blockquote></details>

Done.

---

*[math/test/rotation_conversion_test.cc, line 360 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB4smN72f1ZGfrlkxx:-LGD3He142d5vSz5CBcR:b-cmjck3) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_conversion_test.cc#L360)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

any reason why these power of two values multiplying `kEpsilon`? maybe worth documenting so I can learn something useful :-).
also, why two orders of magnitude bigger than epsilon? a comment would help.
</blockquote></details>

Changed to decimal 32  (5 bits in the 53 bit mantissa of the 64 bit double).

---

*[math/test/rotation_matrix_test.cc, line 424 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB2MSH5vDAqAQGk5HC:-LGD4y498nu4X_DuVnvb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L424)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, it's merged.
</blockquote></details>

Done.

---

*[math/test/rotation_matrix_test.cc, line 442 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB33zr5EW6CcxRcRWT:-LGD5Jtn-7PrFwX2x3kw:b2v5v1b) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L442)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

any thoughts/comments on why `800`? just from running the test or is there an order of magnitude calculation you are doing?
</blockquote></details>

Done.  Changed to 16.  Good catch.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Feature done. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 7 of 7 files at r3.
Review status: 2 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing

---

*[math/rotation_matrix.h, line 36 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVF5KS4y4JAO4YG8ZE:-LGVF5KT-oS421lFH2x8:b-79vg54) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/rotation_matrix.h#L36)):*
> ```objc
> /// For speed in release builds, all validity checks are skipped except a
> /// validity check is always made (in both release and debug builds) for the
> /// inherently computationally expensive method ProjectToRotationMatrix().
> ```

nit: this kind of comments are usually difficult to maintain in sync with the source. Do you think this will be true forever? maybe there is another way to paraphrase this?

---

*[math/test/rotation_conversion_test.cc, line 360 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB4smN72f1ZGfrlkxx:-LGVFUmUEyMZmv8WFx8s:b-1mi5xy) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_conversion_test.cc#L360)):*
<details><summary><i>Previously, mitiguy (Mitiguy) wrote…</i></summary><blockquote>

Changed to decimal 32  (5 bits in the 53 bit mantissa of the 64 bit double).
</blockquote></details>

so no reason then? I mean, 30 would work just fine right?

---

*[math/test/rotation_matrix_test.cc, line 415 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB1RAW80OoDE2uBnzD:-LGVFb5iDjJXz6OLZoEm:b-qi4eah) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L415)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Yes, it's doable. I'll add it to my test cases later.
</blockquote></details>

nice! that'll be beautiful. 

---

*[math/test/rotation_matrix_test.cc, line 424 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB2MSH5vDAqAQGk5HC:-LGVFhfV-yFweAKhU2yW:bb7l8rz) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L424)):*
<details><summary><i>Previously, mitiguy (Mitiguy) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

neat

---

*[math/test/rotation_matrix_test.cc, line 442 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB33zr5EW6CcxRcRWT:-LGVFlsV4kJQ5n-n7dKI:be1o8i0) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L442)):*
<details><summary><i>Previously, mitiguy (Mitiguy) wrote…</i></summary><blockquote>

Done.  Changed to 16.  Good catch.
</blockquote></details>

nit: space around multiplication operator `*`. Here and maybe other places.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037#-:-LGVFtr10zylFFX3INTA:b-pf8zwv)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[math/test/rotation_matrix_test.cc, line 442 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB33zr5EW6CcxRcRWT:-LGVI9EZ7Y3C9tuaQJZ0:bcixxgz) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_matrix_test.cc#L442)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

nit: space around multiplication operator `*`. Here and maybe other places.
</blockquote></details>

@amcastro-tri GSG permits zero-space around `*` and `/` operations.  http://drake.mit.edu/styleguide/cppguide.html#Horizontal_Whitespace

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
+(status: curate commits before merging) 

First pass done.

---

Reviewed 2 of 7 files at r1, 1 of 6 files at r2, 7 of 7 files at r3.
Review status: 5 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVRnoj5hBBDFFEfFTT:-LGVRnok2IvjkV3OmjIO:b-xiah7i) (no related file):*
FYI @mitiguy You want to give a bit more documentation in the PR to guide your reviewers. The *title* is to allow `RotationMatrix` to be instantiated on symbolic type, but that doesn't lead one to expect to see, `Transform` and `RollPitchYaw` appearing. A bit more documentation so that the reviewer has context to know what has happened and why things are there is always a boon.

---

*[math/rotation_matrix.h, line 34 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVSOZL4xlodkfzveXu:-LGVSOZL4xlodkfzveXv:bik63xz) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/rotation_matrix.h#L34)):*
> ```objc
> /// @note In debug builds, several methods in this class do a validity check
> /// and throw an exception (std::logic_error) if the rotation matrix is invalid.
> /// For speed in release builds, all validity checks are skipped except a
> ```

FYI This sentence got away from you: "*all* validity checks are skipped except a validity check is *always* made..."  Related to note below about code-comments syncrhonization.

---

*[math/rotation_matrix.h, line 36 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVF5KS4y4JAO4YG8ZE:-LGVSitk27nk0pFJQLSE:bmi8a9d) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/rotation_matrix.h#L36)):*
As to @amcastro-tri 's comment, perhaps it would be better to say:

> This class has the ability to validate `RotationMatrix` invariants. The circumstances when these validations are performed depends on whether `DRAKE_ASSERT_IS_ARMED` is defined or not. If defined, the maximum number of validations are performed. When not defined, a minimum number of validations are performed. No validation is ever performed for non-numeric scalar types (e.g., symbolic::Expression).

The points I'd underscore:

1. This means we can arbitrarily change which tests are *always* tests and which tests are *depends* test whenever we want without having to change the documentation.
2. Technically, it's not "in debug builds" that trigger the tests, it's arming drake assert (which can be done in release builds).

---

*[math/rotation_matrix.h, line 69 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVVAVN8w85mTBvS8q2:-LGVVAVN8w85mTBvS8q3:b-kion2k) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/rotation_matrix.h#L69)):*
> ```objc
>   /// Constructs a %RotationMatrix from an Eigen::Quaternion.
>   /// @param[in] quaternion a non-zero, finite quaternion which may or may not
>   /// have unit length [i.e., `quaterion.norm()` does not have to be 1].
> ```

FYI s/quaterion/quaternion

---

*[math/rotation_matrix.h, line 413 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVbzoFBhxUFQSeDN_a:-LGVbzoFBhxUFQSeDN_b:bk3hpvi) ([raw file](https://github.com/robotlocomotion/drake/blob/e13d95847a966ac83eb09dbb0ec86a829be22832/math/rotation_matrix.h#L413)):*
> ```objc
>   /// Engineering and Computer Science, Massachusetts Institute of Technology"
>   /// https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-241j-dynamic-systems-and-control-spring-2011/readings/MIT6_241JS11_chap04.pdf
>   // @internal This function is not generated for symbolic Expression.
> ```

You deleted the statement that it wasn't implemented for symbolic expression, however, it should be documented that while this function exists for all scalar types, invocation on symbolic::Exprsesion (and other, non-numeric scalar types) will, in fact, throw an exception.

---

*[math/rotation_matrix.h, line 439 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVdz_AAejussH0FEFq:-LGVdOGnBEvhCBkEjjcj:b3tgln1) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/rotation_matrix.h#L439)):*
> ```objc
>   ProjectToRotationMatrix(const Matrix3<S>& M, T* quality_factor = NULL) {
>     throw std::runtime_error("This method is not supported for scalar types "
>                              "that are not is_numeric<S>.");
> ```

nit: s/is_numeric<S>/drake::is_numeric<S>

Because it *looks* like it could be a c++ type trait, we want to make sure it's clear that it is a custom type trait.

---

*[math/test/rotation_conversion_test.cc, line 360 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGB4smN72f1ZGfrlkxx:-LGVi3hbC0qFTMh9mYAI:bdrff66) ([raw file](https://github.com/robotlocomotion/drake/blob/71d5e1ba7dcb388b4028ad16bfa4c7f5a1967a97/math/test/rotation_conversion_test.cc#L360)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

so no reason then? I mean, 30 would work just fine right?
</blockquote></details>

Meta: Generally, you have some really specific scale factors on your epsilon. You should have some documentation somewhere that explains why these are "right" or "good". Future developers will look at such magical numbers and wrack their brains trying to infer their significance.

---

*[math/test/rotation_matrix_test.cc, line 429 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVgDPo10O9tptUEzdE:-LGVgDPo10O9tptUEzdF:bk1awmm) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/test/rotation_matrix_test.cc#L429)):*
> ```cpp
>   EXPECT_TRUE(is_identity.value());
> 
>   // Verify Bool method IsExactlyIdentity().
> ```

This test doesn't seem to meaningfully exercise the `symbolic::Expression`. This test has already been done in the `IsExactlyIdentity` test (line 230).

---

*[math/test/rotation_matrix_test.cc, line 462 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVgdXL6HGChorx5YLu:-LGVgdXL6HGChorx5YLv:b-t6tcd7) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/test/rotation_matrix_test.cc#L462)):*
> ```cpp
>   EXPECT_TRUE(is_exactly_equalA.value());
>   EXPECT_FALSE(is_exactly_equalB.value());
> }
> ```

I don't see the disabling of `ThrowIfNotValid` for `symbolic::Expression` tested here. I should see a test that is exercised only if `DRAKE_ASSERT` is armed and some representative set of functions that would invoke `ThrowIfNotValid` should essentially do nothing.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037#-:-LGVh2M09ouvgsxfqGaM:bc2jdps)*
<!-- Sent from Reviewable.io -->

mitiguy commented:
Review status: 3 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing, commit history needs curation

---

*[math/rotation_matrix.h, line 413 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVbzoFBhxUFQSeDN_a:-LGWZp6KCzbdB1l4jDNs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e13d95847a966ac83eb09dbb0ec86a829be22832/math/rotation_matrix.h#L413)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

You deleted the statement that it wasn't implemented for symbolic expression, however, it should be documented that while this function exists for all scalar types, invocation on symbolic::Exprsesion (and other, non-numeric scalar types) will, in fact, throw an exception.
</blockquote></details>

Done.

---

*[math/test/rotation_matrix_test.cc, line 429 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVgDPo10O9tptUEzdE:-LGWtk9X8GmP05Y8Ky12:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/test/rotation_matrix_test.cc#L429)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

This test doesn't seem to meaningfully exercise the `symbolic::Expression`. This test has already been done in the `IsExactlyIdentity` test (line 230).
</blockquote></details>

Done.

---

*[math/test/rotation_matrix_test.cc, line 462 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVgdXL6HGChorx5YLu:-LGWwC3N96spUMOGAnwH:b-juse2w) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/test/rotation_matrix_test.cc#L462)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I don't see the disabling of `ThrowIfNotValid` for `symbolic::Expression` tested here. I should see a test that is exercised only if `DRAKE_ASSERT` is armed and some representative set of functions that would invoke `ThrowIfNotValid` should essentially do nothing.
</blockquote></details>

Done (I think).  PTAL to see if this does what you wanted.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Second pass done.

---

Reviewed 3 of 3 files at r4.
Review status: 6 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing, commit history needs curation

---

*[math/rotation_matrix.h, line 413 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVbzoFBhxUFQSeDN_a:-LGk1gAk4jM_0z7tBX7W:b-nk6plk) ([raw file](https://github.com/robotlocomotion/drake/blob/e13d95847a966ac83eb09dbb0ec86a829be22832/math/rotation_matrix.h#L413)):*
<details><summary><i>Previously, mitiguy (Mitiguy) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I would advocate that this documentation should be part of the public documentation -- callers should know about this behavior.

---

*[math/test/rotation_conversion_test.cc, line 376 at r4](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGk1r7F6cLq66oZh-ao:-LGk1r7F6cLq66oZh-ap:b4iokxn) ([raw file](https://github.com/robotlocomotion/drake/blob/ed958b91fa30f5c850be13114154b51898fe6321/math/test/rotation_conversion_test.cc#L376)):*
> ```cpp
>     // Ensure the calculated quaternion produces the same rotation matrix.
>     // This test accuracy to near machine precision and uses a tolerance of
>     // 32*kEpsilon (allows for 5 of the 53 mantissa bits to be inaccurate).
> ```

BTW This comment deals with "what" the code does. However, any reader can tell that it's close to machine precision and can tell that it loses five bits. That is part of the code. The comment should serve the purpose of saying *why* five bits.

---

*[math/test/rotation_matrix_test.cc, line 429 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGVgDPo10O9tptUEzdE:-LGk43Ji4F98AUvXrhJ6:b-28ma4s) ([raw file](https://github.com/robotlocomotion/drake/blob/a3ad6bf24a3cdb3c87615933c6dd6c60ddb4cb6b/math/test/rotation_matrix_test.cc#L429)):*
<details><summary><i>Previously, mitiguy (Mitiguy) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

My concern obviously isn't clear.

1. This test is called `SymbolicRotationMatrices`
2. This test takes a `Matrix3<double>` and creates a `RotationMatrix<double>` from it.
3. It performs operations on the `RotationMatrix<double>`
4. There is an absolute void of `symbolic::Expression` in this sub test..

I fail to see how these lines of code belong in a test called `SymbolicRotationMatrices`. In contrast, they seem to duplicate the tests I cited above.

---

*[math/test/rotation_matrix_test.cc, line 238 at r4](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGk2Hfy6z7fOpEduS3L:-LGk2Hfy6z7fOpEduS3M:bsz7yqa) ([raw file](https://github.com/robotlocomotion/drake/blob/ed958b91fa30f5c850be13114154b51898fe6321/math/test/rotation_matrix_test.cc#L238)):*
> ```cpp
>        0, 1, 0,
>        0, 0, 1;
>   R.SetOrThrowIfNotValid(m);
> ```

nit: You could/should perhaps wrap this in an `EXPECT_NO_THROW`. Same for the m which is epsilon away from the identity.

---

*[math/test/rotation_matrix_test.cc, line 243 at r4](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGk2T6yArjmtnahV9vd:-LGk2T6yArjmtnahV9ve:b-opci3q) ([raw file](https://github.com/robotlocomotion/drake/blob/ed958b91fa30f5c850be13114154b51898fe6321/math/test/rotation_matrix_test.cc#L243)):*
> ```cpp
>   m(0, 2) = kEpsilon;
>   R.SetOrThrowIfNotValid(m);
>   EXPECT_FALSE(R.IsExactlyIdentity().value());
> ```

btw I *hate* these semantics. A function called `IsIdentity()` can take an epsilon and that sounds good. But the function `IsExactlyidentity()` seems to preclude the idea of tolerances.

I'm not defecting this or calling for a change. I just wanted to announce my disapprobation.

---

*[math/test/rotation_matrix_test.cc, line 250 at r4](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGk3CzW0V8mn9ZsWJTx:-LGk3CzW0V8mn9ZsWJTy:b-6ax1sc) ([raw file](https://github.com/robotlocomotion/drake/blob/ed958b91fa30f5c850be13114154b51898fe6321/math/test/rotation_matrix_test.cc#L250)):*
> ```cpp
>        0, cos_theta, sin_theta,
>        0, -sin_theta, cos_theta;
>   R.SetOrThrowIfNotValid(m);
> ```

You've added a test where you show a small perturbation does not disqualify the matrix from being an identity. It would be good to add another test where a slightly *larger* perturbation precludes the matrix from being considered the identity. These sandwich tests provide strong characterization of the behavior of the tolerance.

You can still leave this test in -- the test where the matrix is arbitrarily distant from the threshold.

---

*[math/test/rotation_matrix_test.cc, line 440 at r4](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGk52uLA-thH5UoRKZ0:-LGk52uLA-thH5UoRKZ1:b-mkm99v) ([raw file](https://github.com/robotlocomotion/drake/blob/ed958b91fa30f5c850be13114154b51898fe6321/math/test/rotation_matrix_test.cc#L440)):*
> ```cpp
> 
>   // Verify Bool methods IsExactlyIdentity(), IsIdentityToInternalTolerance().
>   // Note: Some of these tests are repeated above (OK).
> ```

BTW Why are they repeated? What value does the repetition serve? And what does "OK" mean?

---

*[math/test/rotation_matrix_test.cc, line 447 at r4](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGk5OH9DzgFRPFYJ8gk:-LGk5OH9DzgFRPFYJ8gl:b3gla6t) ([raw file](https://github.com/robotlocomotion/drake/blob/ed958b91fa30f5c850be13114154b51898fe6321/math/test/rotation_matrix_test.cc#L447)):*
> ```cpp
> 
>   // Verify Bool method IsOrthonormal();
>   const Bool<double> is_orthonormal =
> ```

In fact, previous comment applies here and below. Why are tests that have *nothing* to do with `symbolic::Expression` located in a test called `SymbolicRotationMatrices`?

---

*[math/test/rotation_matrix_test.cc, line 475 at r4](https://reviewable.io/reviews/robotlocomotion/drake/9037#-LGk4lqe9vVwe79EHsOf:-LGk4lqe9vVwe79EHsOg:b9ojdiw) ([raw file](https://github.com/robotlocomotion/drake/blob/ed958b91fa30f5c850be13114154b51898fe6321/math/test/rotation_matrix_test.cc#L475)):*
> ```cpp
>   EXPECT_FALSE(is_exactly_equalB.value());
> 
>   // Ensure ThrowIfNotValid() does nothing for symbolic::Expression
> ```

nit: Two notes:

1. Final period.
2. The comment should emphasize the purpose of this test -- executing this test with `DRAKE_ASSERT` disarmed is a *meaningless* test. Only with `DRAKE_ASSERT` armed does this become a meaningful test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9037#-:-LGk5_Ad7fL49Kmd6qG8:bwje0g4)*
<!-- Sent from Reviewable.io -->
